FWCore/Framework/src/EventSetupsController.cc

```diff
@@ -58,29 +71,106 @@ namespace edm {
       finishConfiguration();
-      std::for_each(providers_.begin(), providers_.end(), [&syncValue](std::shared_ptr<EventSetupProvider> const& esp) {
-        esp->eventSetupForInstance(syncValue);
-      });
+
+      bool newEventSetupImpl = false;
+      eventSetupImpls.clear();
+      eventSetupImpls.reserve(providers_.size());
+
+      // Note that unless there are one or more SubProcesses providers_ will only
+      // contain one element.
+
+      for (auto& eventSetupProvider : providers_) {
+        eventSetupProvider->setAllValidityIntervals(syncValue);
+      }
+
+      for (auto& eventSetupRecordIOVQueue : eventSetupRecordIOVQueues_) {
+        // For a particular record, if the top level process or any SubProcess requires
+        // starting a new IOV, then we must start a new IOV for all of them. And we
+        // need to know whether this is needed at this point in time. This is
+        // recorded in the EventSetupRecordProviders.
+        eventSetupRecordIOVQueue->setNewIntervalForAnySubProcess();
+      }
+
+      for (auto& eventSetupProvider : providers_) {
+        // Decides whether we can reuse the existing EventSetupImpl and if we can
+        // returns it. If a new one is needed it will create it, although the pointers
+        // to the EventSetupRecordImpl's will not be set yet in the returned EventSetupImpl
+        // object.
+        eventSetupImpls.push_back(eventSetupProvider->eventSetupForInstance(syncValue, newEventSetupImpl));
+      }
+
+      for (auto& eventSetupRecordIOVQueue : eventSetupRecordIOVQueues_) {
+        eventSetupRecordIOVQueue->checkForNewIOVs(taskToStartAfterIOVInit, endIOVWaitingTasks, newEventSetupImpl);
+      }
     }
 
-    void EventSetupsController::forceCacheClear() const {
-      std::for_each(providers_.begin(), providers_.end(), [](std::shared_ptr<EventSetupProvider> const& esp) {
-        esp->forceCacheClear();
-      });
+    void EventSetupsController::eventSetupForInstance(IOVSyncValue const& syncValue) {
+      // This function only supports use cases where the event setup
+      // system is used without multiple concurrent IOVs.
+      // At the time this comment is being written, this is used for
+      // run transitions and in unit test code. In the future,
+      // it may only be needed for unit tests. This function uses
+      // the other version of eventSetupForInstance that
+      // supports concurrent IOVs. To get this to work, a couple
+      // arguments to that function need dummy objects that do
+      // not serve any purpose in this context. We also need to
+      // add in a task to wait for the asynchronous initialization
+      // of IOVs to complete.
+
+      auto waitUntilIOVInitializationCompletes = make_empty_waiting_task();
+      waitUntilIOVInitializationCompletes->increment_ref_count();
+
+      // These do nothing ...
+      WaitingTaskList dummyWaitingTaskList;
+      std::vector<WaitingTaskHolder> dummyVector;
```

Comment:
This variable does not appear to be used anywhere.

Reply:
Removed it. Thanks for catching that.